src/main/connection_listener/exchange_msg.rs

```diff
@@ -333,10 +325,62 @@ impl<UID: Uid> ExchangeMsg<UID> {
             return self.terminate(core, poll);
         }
 
-        self.enter_handshaking_mode(their_uid);
+        if self.test_ext_reachability {
+            let on_check_reachability_result =
+                |mut state: RefMut<ExchangeMsg<UID>>,
+                 core: &mut EventLoopCore,
+                 poll: &Poll,
+                 child,
+                 res| {
+                    state.handle_check_reachability_on_connect(core, poll, child, res);
+                };
+            self.spawn_ext_reachability_tests(
+                core,
+                poll,
+                their_uid,
+                their_addrs,
+                their_pk,
+                on_check_reachability_result,
+            );
+            if self.reachability_children.is_empty() {
+                trace!("External reachability test failed. Denying connect request.");
+                let reason = BootstrapDenyReason::FailedExternalReachability;
+                self.write(core, poll, Some((Message::BootstrapDenied(reason), 0)));
+            }
+        } else {
+            self.send_connect_grant(core, poll, their_uid);
+        }
+    }
+
+    /// Called when external reachability test passed.
+    fn handle_check_reachability_on_connect(
+        &mut self,
+        core: &mut EventLoopCore,
+        poll: &Poll,
+        child: Token,
+        res: Result<UID, ()>,
+    ) {
+        let _ = self.reachability_children.remove(&child);
+        if let Ok(their_uid) = res {
+            self.terminate_childern(core, poll);
+            return self.send_connect_grant(core, poll, their_uid);
+        }
+        if self.reachability_children.is_empty() {
+            trace!("External reachability test failed, terminating connection.");
```

Comment:
these should probably be debugs now as it's become more important to note these failures.